_posts/2020-04-27-tworzenie-i-usuwanie-indeksow-w-bazie-mongodb-na-produkcji.md

```diff
@@ -0,0 +1,79 @@
+
+Dzięki shardingowi MongoDB pozwala przechowywać i przetwarzać ogromne ilości danych bez znacznego spadku wydajności. Poza tym dzięki redundancji w klastrze (_replica set_) dane są niemalże zawsze dostępne. Co więcej, dokumentacja MongoDB wprost mówi, że klaster z redundacją danych powinien być podstawą każdej instalacji produkcyjnej.
+
+Każdy system jest rozwijany i zmienia się w odpowiedzi na nowe wymagania biznesowe. Z tego powodu może się zdarzyć, że konieczne będą zmiany w zapytaniach do bazy i co za tym idzie również mogą być wymagane zmiany indeksów. Staniemy wtedy przed wyzwaniem zmiany indeksów w systemie produkcyjnym. Może się zdarzyć, że jeden okaże się niepotrzebny i będziemy chcieli go usunąć, aby zwolnić miejsce. Drugi indeks trzeba będzie stworzyć, aby przyspieszyć wykonywanie nowych zapytań. Tworzenie indeksu może trwać bardzo długo. Nie chcemy na ten czas wyłączać produkcji. Poniższy tekst wyjaśnia, jak stworzyć nowy i usunąć stary indeks bez większych zawirowań na produkcji.
+
+Aby cała operacja odbyła się bez problemów, powinniśmy wykonać ją w następującej kolejności:
+- utworzenie indeksu pod nowe zapytania w systemie;
+- aktualizacja części biznesowej systemu;
+- usunięcie starego indeksu, o ile nie jest już używany.
+
+# Tworzenie indeksu w MongoDB w sposób _rolling_
+
+Tworzenie indeksu w sposób _rolling_ dotyczy tylko i wyłącznie bazy z redundacją danych w klastrze. Ta metoda pozwala na tworzenie unikalnych indeksów. Przeprowadzana jest dla każdego węzła w klastrze:
+- węzeł jest odłączany od klastra i uruchamiany w trybie standalone;
+- tworzony jest indeks na tym węźle;
+- węzeł jest dołączany do klastra.
+
+W czasie tworzenia indeksu pozostałe węzły klastra cały czas działają produkcyjnie, i co z tym idzie dane się zmieniają. Węzeł, który był uruchomiony w trybie standalone, i na którym tworzony był indeks, po włączeniu do klastra będzie musiał się zsynchronizować i pobrać wszytkie zmiany, które zaszły w czasie, kiedy był odłączony. Aby cała procedura odbyła się bez problemu, należy zapewnić, aby oplog miał odpowiedni rozmiar. Oplog przechowuje ostatnie zmiany danych, które zaszły w węźle primary i jeżeli nie będzie odpowiednio duży, to węzeł który jest dołączany do klastra, nie będzie w stanie się zsynchronizować. Więcej informacji o tym, jak dobrać rozmiar oplog jest w [dokumentacji MongoDB](https://docs.mongodb.com/manual/core/replica-set-oplog/#replica-set-oplog-sizing).
+
+Procedurę zaczynamy od węzłów secondary, a na końcu przeprowadzamy ją dla węzła primary.
```

Comment:
Dopisałbym tutaj jeszcze paragraf o Write Concern. Jeśli kaster składa się z 3 węzłów i ustawimy w:3, no to system nie będzie dostępny. Wszystkie zapisy będą się kończyć błędem.

Reply:
Dopisałem